### PR TITLE
Test model

### PR DESCRIPTION
**Pull request by Evesonr**
Adding a script to run a multiclass classification model on the 20newsgroups dataset.

### CODE REVIEW

File: model.py [hunk 1]
```diff
@@ -0,0 +1,48 @@
+
+
+import numpy as np
+from numpy.linalg import inv
```
Comment (Evesonr) [inline model.py:4]: You can reduce the number of imports you're making here. For example, use `import numpy as np` and call `np.linalg.inv` in the code below. This way, you don't need to import inv from numpy.linalg here, particularly as you are only using inv once.

File: model.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
+
+
+import numpy as np
+from numpy.linalg import inv
+
+from sklearn import datasets 
+from sklearn.decomposition import PCA
```
Comment (Evesonr) [inline model.py:7]: I notice you import PCA (good thinking!), but you don't use it. PCA could be an option for dimensionality reduction with a modelling task like this, but in general it is best to remove any unused imports in your script to improve efficiency and legibility. Remember - code is written once and read 100s of times :)

File: model.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
+
+
+import numpy as np
+from numpy.linalg import inv
+
+from sklearn import datasets 
+from sklearn.decomposition import PCA
+from sklearn.feature_extraction. text import TfidfVectorizer 
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import Normalizer, StandardScaler
```
Comment (Evesonr) [inline model.py:10]: StandardScaler - see my comment on imports and PCA above.

File: model.py [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+import numpy as np
+from numpy.linalg import inv
+
+from sklearn import datasets 
+from sklearn.decomposition import PCA
+from sklearn.feature_extraction. text import TfidfVectorizer 
+from sklearn.pipeline import make_pipeline
+from sklearn.preprocessing import Normalizer, StandardScaler
+
+class LinearRegression:
```
Comment (Evesonr) [inline model.py:12]: From your email, I understand you're trying to identify topics in news articles? This is a multiclass classification task - we are trying to model which topic or topics a given observation pertains to, out of a set of known topics.

However, linear regression is not suitable for this task - linear regression is more suited to when we want to predict a value on a continuous scale (e.g. we have the height of a sample of people, and want to predict their weight).

A good choice of model here might be logistic regression, or other models such as naive Bayes or a random forest classifier. These models are more suited to classification tasks - they return a discrete output. You can read more [here](https://www.analyticsvidhya.com/blog/2020/12/beginners-take-how-logistic-regression-is-related-to-linear-regression/#:~:text=Linear%20regression%20provides%20a%20continuous,values%20to%20the%20sigmoid%20curve.)

File: model.py [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+    @staticmethod
+    def loss (predictions, y):
+        return ((predictions - y) ** 2).sum() / len(y)
```
Comment (Evesonr) [inline model.py:15]: It's great that you can define the mean squared error here. However, if we are using a different algorithm, say logistic regression, we would use a different loss, such as cross entropy loss.

Also, a common way to calculate this loss would be to use the 'out of the box' functions in scikit-learn, as these have been optimised for performance and can reduce the lines of code you may need to write. You could import log_loss from sklearn.metrics if you wish to do this. Before doing so, please see my comment in email about using scikit-learn models.

File: model.py [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+        return self
+        
+    def predict(self, X):
+        return (X @ self.weights).sum(axis=1)
+
+
+def train_model(X,Y):
+    model = make_pipeline(
+        TfidfVectorizer(analyzer="char"),
+        Normalizer(),
```
Comment (Evesonr) [inline model.py:30]: I believe TfidfVectorizer returns normalised vectors by default. So you could remove this additional Normalizer step in the pipeline :)

File: model.py [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+    predictions = model.predict(X) 
+    return LinearRegression.loss(predictions, y)
+
+
+newsgroup_data = datasets.fetch_20newsgroups(subset="train")
+X = newsgroup_data.data
+y = newsgroup_data.target
+
+model = train_model(X, y)
+test_result = test_model(model, X, y)
```
Comment (Evesonr) [inline model.py:46]: Good use of the sklearn.datasets module. However, it's really important to also download the test set, as well as the training set. Currently you are training the model, and evaluating it on the same data. If your model has learned from this data, the test_model function is likely to return very good results on the train set that are unlikely to be replicated when you run your model on new data, i.e. a biased evaluation. To provide an unbiased evaluation of your trained model, you should run your test_model function on a set of unseen data. The easiest way to do that is to run it on the test set from 20newsgroups. See the [sklearn docs ](https://scikit-learn.org/stable/modules/generated/sklearn.datasets.fetch_20newsgroups.html)for more info.

File: model.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+        TfidfVectorizer(analyzer="char"),
+        Normalizer(),
+        LinearRegression()
+    )
+    model.fit(X, y) 
+    return model
+
+def test_model(model, X, y):
+    predictions = model.predict(X) 
+    return LinearRegression.loss(predictions, y)
```
Comment (Evesonr) [inline model.py:38]: Consider returning evaluation metrics too, for example accuracy, F1 score, precision and recall. These will be useful when testing the model to understand it's performance.

File: model.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        return ((predictions - y) ** 2).sum() / len(y)
+
+    def fit(self, X, y):
+        n_features = np.size(X, 1)
+        y = np.expand_dims(y, axis=1)
+        self.weights = inv(X.T @ X) @ X.T @ y 
+        return self
+        
+    def predict(self, X):
+        return (X @ self.weights).sum(axis=1)
```
Comment (Evesonr) [inline model.py:24]: I believe you are summing the class probabilities across each row here. If this is a multiclass classification problem, it may be more effective to simply return the probabilities by class - i.e. each row has the predicted probabilities for each class. We can then use these probabilities to calculate the loss, and to infer the highest-probability class for each row.

**Comment by Evesonr** [comment]
A great start! Just a few things to iron out. If you'd like to chat through any of my suggestions please let me know.
